target/linux/ath79/image/generic-tp-link.mk

```diff
@@ -362,6 +362,17 @@ define Device/tplink_cpe610-v2
 endef
 TARGET_DEVICES += tplink_cpe610-v2
 
+define Device/tplink_cpe710-v1
+  $(Device/tplink-safeloader-uimage)
+  SOC := qca9563
+  IMAGE_SIZE := 15680k
```

Comment:
Should this be: firmware size 0xf40000 / 1024 = 15616k ?

Reply:
@jwmullally The Firmware size is actually 0xf50000 which is 15680k
Refer to the DTS target/linux/ath79/dts/qca9563_tplink_cpe710-v1.dts

Reply:
Is there a mismatch between the DTS and tplink-safeloader.c partition list?
```
			{"device-id", 0x63000, 0x00100},
			{"firmware", 0x70000, 0xf40000},
			{"soft-version", 0xfb0000, 0x00100},
			{"support-list", 0xfb1000, 0x01000},
			{"user-config", 0xfc0000, 0x10000},
```

```
			info: partition@60000 {
				label = "info";
				reg = <0x060000 0x010000>;
				read-only;
			};
			partition@70000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x070000 0xf50000>;
			};
			partition@fc0000 {
				label = "config";
				reg = <0xfc0000 0x030000>;
				read-only;
			};
```

Reply:
@jwmullally, see the discussion at https://github.com/openwrt/openwrt/pull/4143#issuecomment-841177042
I believe this mismatch is intentional, and used to reclaim some flash space for use by the rootfs partition.

Reply:
Rearranging the non-persistent partitions is fine.

Having `firmware` have different sizes in tplink-safeloader and the DTS/Makefile just to overlap `soft-version` and `support-list` like this is basically taking advantage of edge-case undefined behaviour in tplink-safeloader, TP-Link's OEM firmware upgrade and sysupgrade to get 64k (0.4%) more space. I don't think its worth the risk, and will just result in ongoing confusion and bugs. Has it been tested to the full firmware size?

I think its much better to stick with a standard predictable partition layout where nothing overlaps and everything is aligned. 

Either (preferable):
```
# tplink-safeloader.c
			{"firmware", 0x70000, 0xf40000},
			{"soft-version", 0xfb0000, 0x00100},
			{"support-list", 0xfb1000, 0x01000},
			{"user-config", 0xfc0000, 0x10000},
# DTS
			partition@70000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x070000 0xf50000>;
			};
			partition@fc0000 {
				label = "config";
				reg = <0xfc0000 0x030000>;
				read-only;
			};
```
or much messier (might not be possible to due to alignment issues? Not worth it for 59k imho):
```
# tplink-safeloader.c
			{"firmware", 0x70000, 0xf4ef00},
			{"soft-version", 0xfbef00, 0x00100},
			{"support-list", 0xfbf000, 0x01000},
			{"user-config", 0xfc0000, 0x10000},
# DTS
			partition@70000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x070000 0xf4ef00>;
			};
			partition@fc0000 {
				label = "config";
				reg = <0xfbf000 0x031100>;
				read-only;
			};
```

If this overlap is being kept, the firmware section of the tplink-safeloader.c and DTS should include comments explaining whats going on.

Reply:
@svanheule Ok, I see now this is historically how it works for *all* the 15 existing CPE/WBS devices, and none of the other 46 TP-Link units, so this discussion probably belongs in another PR like [the one you linked above](https://github.com/openwrt/openwrt/pull/4143).

Okay, you can ignore this particular thread, sorry for the distraction :)

```
$ grep "\.id\|\.last_sysupgrade_partition" tools/firmware-utils/src/tplink-safeloader.c 
		.id     = "CPE210",
		.last_sysupgrade_partition = "support-list",
		.id     = "CPE210V2",
		.last_sysupgrade_partition = "support-list",
...
		.id     = "RE500-V1",
		.last_sysupgrade_partition = "file-system"
		.id     = "RE650-V1",
		.last_sysupgrade_partition = "file-system"
```

Reply:
> For the CPE and WBS devices sysupgrade back to the OEM Firmware generally does not work.

FYI, for this unit, I tested removing the `firmware` DTS -> `soft-version` `support-list` +64k hack (i.e. DTS firmware 0xf40000) and sysupgrade using OEM firmware works again.

Its up to ye, but I would recommend using a non-overlapping "standard" TP-Link partition layout for this unit (i.e. no hack) to put this in "right", and explore separately if this fixes the issue for the other CPE units.

Reply:
As mentioned in https://github.com/openwrt/openwrt/pull/4117 the runtime partition layout may not be identical to the one stored in the partition table. On the other CPE units, the order of the partitions was also modified. Maybe that's the reason sysupgrade-to-stock is broken there, because an incompatible partition layout is used?

Reply:
@svanheule Yes, due to fixed `mtdparts`, OEM sysupgrade may not be guaranteed to work (e..g rootfs would be misaligned by 64k), although it does actually work in this case.
I agree that TFTP should still be the only recommended way of flashing, with OEM sysupgrade as an unsupported nice-to-have expert feature (if it fails, they can just fall back to TFTP). In this case, it could save people taking their unit down from the antennae.

(There a many improvements that could be made to `tplink-safeloader -z oem.bin`. `convert_firmware()` is currently very simple and produces only a raw contiguous output of `os-image`  + `file-system`, without considering if there is anything is in between. It doesn't add safety metadata. It could use an md5sum whitelist of tested OEM firmwares etc).

The overall point of this thread though is just simplifying this area (and not blocking oem sysupgrade) by removing the overlap for 16MB CPE/WBS units.